### DISCUSSION TITLE
New Godot Stage : Linear tracks w/ looming

### DISCUSSION THREAD

**Discussion by ryojimo** [sn-lab/MouseGoggles#26] [Q&A] [answered]
Hello!
I also create a treadmill for simple 1D VR tasks like "[Spherical Treadmill: air cushion vs. ball transfer bearings](https://github.com/sn-lab/MouseGoggles/discussions/16)".
And then, I am training mice on this treadmill using "Linear tracks" stage mainly.
However, this mice's reaction is week.
May I ask to create a new stage which has loom on Linear tracks, if possible ?

**Comment by misaacson01** [comment]
Just to clarify, are you looking to have a new godot experiment (scene + script) which is otherwise identical to the linear track scene, but that it has looming stimuli? If so, yes that is possible, I'd be happy to make that.

**Comment by ryojimo** [comment]
Thank you!! Yes, as you mentioned.
If there is something function to start "looming stimuli" like "start_looming()", it's helpful to customize something.

For example, 

```
func _input(ev):
	if ev is InputEventKey and ev.is_pressed():
		if ev.scancode == KEY_ESCAPE:
			get_tree().quit() 
		if ev.scancode == KEY_DOWN:
			start_looming()

func start_looming():
	# start "looming stimuli".
```

**Reply by misaacson01** [reply, marked as answer]
I think I understand what you're asking for. To get into the details: I don't think a typical function like "func start_looming():" will work, because functions typically block other things from happening until they finish, and looms requires display updates. Maybe a "thread" would work, but then it's getting more complicated, and I'm trying to keep experiments simple so others can more easily customize them.

But what I can do is keep the "loom" code inside the game loop as usual, combine the linear track and loom experiments into one, and give the option to initiate looms with a keypress (instead of timed trials). The key I chose is "L", for loom. I have just uploaded this new experiment "linearTrackLoom" in Godot project V1.8, in the [Godot folder](https://github.com/sn-lab/MouseGoggles/tree/main/Godot). For this project, I'm using MouseGoggles Duo V1.5, so if you're using an older version of the headset you'll have to switch back to that (i.e. using the "headkinbody_headset" nodes in each experiment .tscn file, instead of the "headkinbody_headset_waveshare" which is currently being used). 

Let me know if this answers your question or if anything isn't clear!

**Comment by ryojimo** [comment]
I really appreciate your quick action!
The implementation using flag ( `loom_started` ) is more than enough for my request.
I tried to understand how "looms" works in V1.6 at first, but I couldn't.
However, I think to be able to understand it by V1.8.
I think i'll try to do a bit of customization.
For example...

- Trigger it, when mice moves 1 meter from the starting-position.
- Trigger it, every 10 minutes.
- etc


I use "Waveshare 19192([link](https://www.waveshare.com/product/1.28inch-lcd-module.htm))" and V1.8 looks to work well.

```
func _input(ev):
	if ev is InputEventKey and ev.is_pressed():
		if ev.scancode == KEY_ESCAPE:
			get_tree().quit() 
		if ev.scancode == KEY_L:
			loom_started = true

```
